src/main/com/fulcrologic/fulcro/algorithms/normalized_state_helpers.cljc

```diff
@@ -0,0 +1,340 @@
+          (dissoc m k)))
+      m)
+    (dissoc m k)))
+
+
+;;====
+
+(>defn remove-entity*
+  "Remove the given entity at the given ident. Also scans all tables and removes any to-one or to-many idents that are
+   found that match `ident` (removes dangling pointers to the removed entity).
+
+   The optional `cascade` parameter is a set of keywords that represent edges that should cause recursive deletes
+   (i.e. it indicates edge names that *own* something, indicating it is safe to remove those entities as well).
+
+   Returns the new state map with the entity(ies) removed."
+
+  ;;TODO
+  ([state-map ident]
+   [map? eql/ident? => map?]
+
+   )
+
+  ;;TODO
+  ([state-map ident cascade]
+   [map? eql/ident? (s/coll-of keyword? :kind set?) => map?]
+   ))
+
+
+
+
+
+(comment
+
+  (def denorm-data {:a [[:b 1]] :b [:b 1]})
+
+  (def original-state
+    {:a      [:b 1]
+     :top-tm [[:d 1] [:b 1]]
+     :denorm {:level-1 {:level-2 denorm-data}}
+     :b      {1 {:c [:d 1]
+                 :e [[:b 1] [:d 1]]}}
+     :d      {1 {:value 42}}})
+
+
+  ;;;;;;;;;;;;;;;;;
+  ;; REFACTOR this section into the proper `remove-entity*` API
+
+
+
+  (defn fun [state-map [a-path a-value]]
+    ;; FIXME take the ident from arguments
+    (let [ident [:b 1]
+          vector-of-vectors? (if (and
+                                   (vector? a-value)
+                                   (every? vector? a-value))
+                               true
+                               false)]
+      (cond
+        ;; nil => dissoc from map
+        (nil? a-value) (dissoc-in state-map a-path)
+        ;; to-many => update the path
+        vector-of-vectors? (assoc-in state-map a-path (apply vector (remove #{ident} a-value)))
+        ;;to-one => remove from the state-map
+        (vector? a-value) (if (= ident a-value)
+                            (dissoc-in state-map a-path)))))
+
+
+  (let [ident [:b 1]
+        nil-or-vector? (fn [x] (or (nil? x)
+                                   (vector? x)))
+        state-after-ident-dissoc (-> original-state (dissoc-in ident))
+        all-paths-after-dissoc-and-denorm-keys (filter
+                                                 (fn [a-path]
+                                                   (if (< (count a-path) 4)
+                                                     true
+                                                     false))
+                                                 (paths state-after-ident-dissoc))
+        path-values-map    (into {}
+                                 (filter #(nil-or-vector? (second %))
+                                         (zipmap
+                                           all-paths-after-dissoc-and-denorm-keys
+                                           (map #(get-in state-after-ident-dissoc %)
+                                                all-paths-after-dissoc-and-denorm-keys))))]
+    (reduce fun
+            state-after-ident-dissoc
+            path-values-map))
+
```

Comment:
@awkay , I need a bit of stylistic advice from you to refactor this solution into the required API. It works fine, however, it doesn't look like even I'd understand this without an effort after a while. I tried many approaches to make this as idiomatic and correct as possible but it doesn't look nice at all.

Reply:
I think, the `cascading` option could also be incorporated via this general approach.

Reply:
This is all part of learning functional decomposition.  Think in terms of naming things for the (simple and single) action that they take. Functions are verbs. Think of the "leaf steps" you need, along with what data needs to be passed to it for it to accomplish the job.  I might have time later to write out some of it as an example.